bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java

```diff
@@ -1680,23 +1016,30 @@ private void updateButtonState() {
 	 * @since 3.0
 	 */
 	private void updateButtonState(boolean disableReplace) {
+		setupFindReplacer();
 		if (okToUse(getShell()) && okToUse(fFindNextButton)) {
 
-			boolean selection= false;
-			if (fTarget != null)
-				selection= !fTarget.getSelectionText().isEmpty();
-
-			boolean enable= fTarget != null && (fActiveShell == fParentShell || fActiveShell == getShell());
-			String str= getFindString();
-			boolean findString= str != null && !str.isEmpty();
+			boolean hasActiveSelection = false;
+			String selection = getCurrentSelection();
+			if (selection != null)
+				hasActiveSelection = !selection.isEmpty();
 
-			fWholeWordCheckBox.setEnabled(isWord(str) && !isRegExSearchAvailableAndChecked());
+			boolean enable = (findReplaceLogic.getTarget() != null)
+					&& (fActiveShell == fParentShell || fActiveShell == getShell());
+			String str = getFindString();
+			boolean findString = str != null && !str.isEmpty();
 
+			fWholeWordCheckBox.setEnabled(isWord(str) && !findReplaceLogic.isRegExSearchAvailableAndActive());
 			fFindNextButton.setEnabled(enable && findString);
-			fSelectAllButton.setEnabled(enable && findString && fTarget instanceof IFindReplaceTargetExtension4);
-			fReplaceSelectionButton.setEnabled(!disableReplace && enable && isEditable() && selection && (!fNeedsInitialFindBeforeReplace || !isRegExSearchAvailableAndChecked()));
-			fReplaceFindButton.setEnabled(!disableReplace && enable && isEditable() && findString && selection && (!fNeedsInitialFindBeforeReplace || !isRegExSearchAvailableAndChecked()));
-			fReplaceAllButton.setEnabled(enable && isEditable() && findString);
+			fSelectAllButton.setEnabled(
+					enable && findString && (findReplaceLogic.getTarget() instanceof IFindReplaceTargetExtension4));
+			fReplaceSelectionButton.setEnabled(!disableReplace && enable && findReplaceLogic.getTarget().isEditable()
+					&& hasActiveSelection && (findReplaceLogic.getTarget().getSelection().toString() != "" //$NON-NLS-1$
+							|| !findReplaceLogic.isRegExSearchAvailableAndActive()));
+			fReplaceFindButton.setEnabled(!disableReplace && enable && findReplaceLogic.getTarget().isEditable()
+					&& findString && hasActiveSelection && (findReplaceLogic.getTarget().getSelection().toString() != "" //$NON-NLS-1$
+							|| !findReplaceLogic.isRegExSearchAvailableAndActive()));
+			fReplaceAllButton.setEnabled(enable && findReplaceLogic.getTarget().isEditable() && findString);
```

Comment:
findReplaceLogic.getTarget() may return null, thus throwing an NPE

Reply:
[Here](https://github.com/eclipse-platform/eclipse.platform.ui/pull/1132/files/3f626d7fd7a603b3f6fd0d0d72e16050eeda0a81#diff-726222a75f354e65760b3be3549d8b994b00aed8e8919328c13560578d893427R1027), I define a boolean expression that will always be evaluated before `findReplaceLogic.getTarget()`. Since it is chained by conjunction, using [short-circuit evaluation](https://docs.oracle.com/cd/E19253-01/817-6223/chp-typeopexpr-6/index.html), I skip the potential NPE.

This has a few advantages and a few drawbacks:
- I need to set an enabled-value for the buttons, and I need to set it (... to disabled) even if the target is not available. Hence, using the method saves repeat-code that would appear by setting the enabled-value in two instances.
- On the other hand, this code is really bad to read. I think that it's fine since it's encapsulated into a well-named small function and I don't think that splitting this up into variables or functions would greatly help with readability. One could easily be misled to - for example - think that there is an NPE occurring.

I have revised the code, extracting findReplaceLogic.getTarget() into a variable and adding a comment that tells future reviewers why this is not an NPE. I'm not happy with that solution, I still think these lines of code are clumsy.
Do you have a better Idea?

Reply:
The NPE you were experiencing came from another source (I had fixed that issue, but I do not remember which and since all commits are squashed, I cannot see the history of where this happened. I will perform extra-many tests to make sure that I really DID address the issue)

Reply:
Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/1d2e0ec38eeb85b31cf30884a622e0e66485ad27